substrate/frame/staking/src/lib.rs

```diff
@@ -1328,18 +1355,97 @@ impl<T: Config, const DISABLING_LIMIT_FACTOR: usize> DisablingStrategy<T>
 		}
 
 		let offender_idx = if let Some(idx) = active_set.iter().position(|i| i == offender_stash) {
 			idx as u32
 		} else {
 			log!(debug, "Won't disable: offender not in active set",);
-			return None
+			return DisablingDecision { disable: None, reenable: None }
 		};
 
 		log!(debug, "Will disable {:?}", offender_idx);
 
-		Some(offender_idx)
+		DisablingDecision { disable: Some(offender_idx), reenable: None }
+	}
+}
+
+/// Implementation of [`DisablingStrategy`] which disables validators from the active set up to a
+/// limit and if the limit is reached and the new offender is higher (bigger punishment/severity)
+/// then it re-enables to lowest offender to free up space for the new offender.
+/// 
+/// An extension of [`UpToLimitDisablingStrategy`].
+pub struct UpToLimitWithReEnablingDisablingStrategy<const DISABLING_LIMIT_FACTOR: usize = 3>;
+
+impl<const DISABLING_LIMIT_FACTOR: usize> UpToLimitWithReEnablingDisablingStrategy<DISABLING_LIMIT_FACTOR> {
+	/// Disabling limit calculated from the total number of validators in the active set. When
+	/// reached re-enabling logic might kick in.
+	pub fn disable_limit(validators_len: usize) -> usize {
+		validators_len
+			.saturating_sub(1)
+			.checked_div(DISABLING_LIMIT_FACTOR)
+			.unwrap_or_else(|| {
+				defensive!("DISABLING_LIMIT_FACTOR should not be 0");
+				0
+			})
+	}
+}
+
+impl<T: Config, const DISABLING_LIMIT_FACTOR: usize> DisablingStrategy<T>
+	for UpToLimitWithReEnablingDisablingStrategy<DISABLING_LIMIT_FACTOR>
+{
+	fn decision(
+		offender_stash: &T::AccountId,
+		offender_slash_severity: Perbill,
+		slash_era: EraIndex,
+		currently_disabled: &Vec<(u32, Perbill)>,
+	) -> DisablingDecision {
+		let active_set = T::SessionInterface::validators();
+
+		// We don't disable for offences in previous eras
+		if ActiveEra::<T>::get().map(|e| e.index).unwrap_or_default() > slash_era {
+			log!(
+				debug,
+				"Won't disable: current_era {:?} > slash_era {:?}",
+				Pallet::<T>::current_era().unwrap_or_default(),
+				slash_era
+			);
+			return DisablingDecision { disable: None, reenable: None }
+		}
+
+		// We don't disable validators that are not in the active set
+		let offender_idx = if let Some(idx) = active_set.iter().position(|i| i == offender_stash) {
+			idx as u32
+		} else {
+			log!(debug, "Won't disable: offender not in active set",);
+			return DisablingDecision { disable: None, reenable: None }
+		};
+
+		// We don't disable more than the limit (but we can re-enable a smaller offender to make space)
+		if currently_disabled.len() >= Self::disable_limit(active_set.len()) {
+			log!(
+				debug,
+				"Reached disabling limit {:?}, checking for re-enabling",
+				Self::disable_limit(active_set.len())
+			);
+
+			// Find the smallest offender to re-enable that is not higher than offender_slash_severity
+			if let Some((smallest_idx, _)) = currently_disabled
+				.iter()
+				.filter(|(_, perbill)| *perbill <= offender_slash_severity)
```

Comment:
Just to clarify - even if there is no offender with a smaller offence we will re-enable the first guy with an offence equal to the new offender. Which is a good thing.

However I think we are biased here towards disabling the validator with the smallest index, which is wrong. If there are multiple elements with the same min value `min_by_key` returns the first one. If I am not mistaken `DisabledValidators` are sorted by key ([here](https://github.com/paritytech/polkadot-sdk/pull/5724/files#diff-8dd07463cc135387ca715a01d60b347276b1ff0b5e73db9dc4e8a1390fb462ceR2303)) so this way we will always reenable the guy with the smallest index.

It's a corner case but it would be nice to handle it if it is not too complicated.

Reply:
You are definitely right. I simply don't think this is anything to worry about. This will only kick in and matter only when we have offender of the same level of severity. Slashes will be dispensed and the newest offender will be disabled. From our perspective as long as lowest severity is re-enabled this is the lowest possible risk. 

We will mention it to SRLabs on audit and see what the think. We can remedy as you mentioned with a random selection for a given severity level instead of arbitrary.